notebooks/platform/xarray/lib/io/codecs_test.py

```diff
@@ -0,0 +1,70 @@
+from typing import Optional
+
+import numpy as np
+
+import pytest
+from hypothesis import given
+from hypothesis.extra.numpy import arrays
+from hypothesis.strategies import integers, tuples
+
+from .codecs import PackGeneticBits
+
+array_t = np.ndarray
+
+
+@pytest.fixture(scope="session")
+def codec() -> PackGeneticBits:
+    return PackGeneticBits()
+
+
+def validate_roundtrip(
+    codec: PackGeneticBits, x: array_t, o: Optional[array_t]
+) -> None:
+    o = codec.decode(codec.encode(x), o)
+    assert o.shape == x.shape
+    assert np.allclose(o, x)
+
+
+# NOTE: strategy for array shapes, positive tuple, max 42 is arbitrary
+twoD_shaped_arrays = tuples(
+    integers(min_value=1, max_value=42), integers(min_value=1, max_value=42)
+)
+
+
+def test_cant_cast_to_int_via_output(codec: PackGeneticBits) -> None:
+    x = np.array([[0, 1], [1, 0]], dtype=np.int)
+    # this works (note that the output is a vector):
+    o = codec.decode(codec.encode(x))
+    assert np.allclose(o, np.array([0, 1, 1, 0]))
+    # but this doesn't (when we try to preserve the shape via output array):
+    with pytest.raises(ValueError, match="When changing to a larger dtype"):
+        x = np.array([[0, 1], [1, 0]], dtype=np.int)
+        validate_roundtrip(codec, x, np.zeros_like(x))
+
+
+def test_simple_encoding(codec: PackGeneticBits) -> None:
+    assert np.allclose(codec.encode(np.array([[1]])), np.array([6, 128]))
```

Comment:
FWIW I found the numpy.testing assert statements to be pretty informative in the event of failure vs just getting a False from `allclose` (cf. https://github.com/pytest-dev/pytest/issues/5347)

Reply:
@eric-czech 👍 updated